text/1012-range-types.rst

```diff
@@ -0,0 +1,277 @@
+
+  function multirange(
+      ranges: set of range<anypoint>
+  ) -> multirange<anypoint>
+
+The constructor must normalize the ranges to be non-overlapping and order them
+based on boundaries. This will ensure that when a ``multirange`` is unpacked
+into its component ranges, the order of the ``range`` values would be based on
+their boundaries rather than arbitrary.
+
+
+Accessing range components
+--------------------------
+
+Several functions will be used to access ``range`` components::
+
+  function get_range_upper(range<anypoint>) -> optional anypoint
+  function get_range_lower(range<anypoint>) -> optional anypoint
+  function is_inclusive_lower(range<anypoint>) -> bool
+  function is_inclusive_upper(range<anypoint>) -> bool
+
+
+Accessing multirange components
+-------------------------------
+
+There will be a function for extracting sub-ranges from a multirange::
+
+  function get_subranges(multirange<anypoint>) -> set of range<anypoint>
+
+By default this function will return the ranges ordered by boundaries
+fromlowest to highest. This is the order of definition and Postgres by default
+returns the sub-intervals in that order.
+
+
+Helper functions
+----------------
+
+In addition to the functions mentioned above there must be a way to generate
+an ordered set of values from a ``range`` or ``multirange``::
+
+  function range_unpack(
+      val: range<anydiscrete>,
+      step: anydiscrete = DEFAULT_STEP
+  ) -> set of anydiscrete
+
+  function range_unpack(
+      val: multirange<anydiscrete>,
+      step: anydiscrete = DEFAULT_STEP
+  ) -> set of anydiscrete
+
+
+  function range_unpack(
+      val: range<anycontiguous>,
+      step: anycontiguous
+  ) -> set of anycontiguous
+
+  function range_unpack(
+      val: multirange<anycontiguous>,
+      step: anycontiguous
+  ) -> set of anycontiguous
+
+The ``range_unpack`` function for ``anydiscrete`` intervals **must** have a
+default (minimal) value for the ``step``. On the other hand ``anycontiguous``
+intervals **must** specify the ``step`` explicitly in all cases as no
+"natural" step value can be asumed. When invoked on an unbounded interval
+``range_unpack`` will raise an error.
+
+We also need functions for determining overlapping and whether something is
+contained in a range::
+
+  function contains(
+      haystack: range<anypoint>,
+      needle: range<anypoint>
+  ) -> bool
+
+  function contains(
+      haystack: range<anypoint>,
+      needle: anypoint
+  ) -> bool
+
+  function contains(
+      haystack: multirange<anypoint>,
+      needle: multirange<anypoint>
+  ) -> bool
+
+  function contains(
+      haystack: multirange<anypoint>,
+      needle: anypoint
+  ) -> bool
+
+  function overlaps(
+      haystack: range<anypoint>,
+      needle: range<anypoint>
+  ) -> bool
+
+  function overlaps(
+      haystack: multirange<anypoint>,
+      needle: multirange<anypoint>
+  ) -> bool
+
+It should be noted that if ``contains(A, B) = true`` then ``overlaps(A, B) =
```

Comment:
Postgres defines a bunch more functions for ranges, such as relative disposition (e.g. "strictly to the left of", "strictly to the right of", etc.).  I think we should reflect those as well.

Reply:
Hmm, yeah I missed an important bit here. We need to define `>`, `>=`, `<`, and `<=` for everything in EdgeQL (because everything must be orderable). So I'll add that to "operators". Meanwhile we'll have functions for "strictly to the left/right" and "does not extend to the left/right of".

The question is what to call these? They are wordy and I'd like to avoid full sentences as function names.